libsolidity/formal/CHC.cpp

```diff
@@ -1622,7 +1623,25 @@ map<unsigned, vector<unsigned>> CHC::summaryCalls(CHCSolverInterface::CexGraph c
 	map<unsigned, vector<unsigned>> calls;
 
 	auto compare = [&](unsigned _a, unsigned _b) {
-		return _graph.nodes.at(_a).name > _graph.nodes.at(_b).name;
+		auto const& a = _graph.nodes.at(_a).name;
+		auto const& b = _graph.nodes.at(_b).name;
+		auto extract = [&](auto const& _s) {
+			int start = 0;
+			if (_s.rfind("summary_", 0) == 0)
+				start = 8;
+			else if (_s.rfind("nondet_call_", 0) == 0)
+				start = 12;
+			else
+				return 0;
+			auto u = string(begin(_s) + start, end(_s)).find('_');
+			solAssert(u != string::npos, "New \"_\" should be present after predicate id.");
+			int result;
+			auto [p, ec] = std::from_chars(_s.data() + start, _s.data() + u, result);
+			solAssert(ec == std::errc(), "Predicate id should be a number.");
+			return result;
+		};
```

Comment:
Would a regexp be easier? something like `(summary_|non_det_call_)(\d*)` and parse the value of the digit?

Reply:
A regex is significantly heavier, isn't it? But would be fine with me, too. But might just
```
auto beg = _s.begin();
while(beg != _s.end() && !isdigit(*beg)) ++beg;
auto end = beg;
while (end != _s.end() && isdigit(*end)) ++end;

solAssert(beg != end, "Expected to find numerical predicate id.");


int result;
auto [p, ec] = std::from_chars(beg, end, result);
solAssert(ec == std::errc(), "Predicate id should be a number.");
return result;
```
be simpler? Is it important what kind of prefix the thing has?

Reply:
@ekpyron By heavier, do you mean performance wise? I would prefer regex over doing stuff on iterators.

Reply:
Yes. Regex in comparison is an extremely slow and complex operation.

Reply:
I agree that regex sounds a bit like overkill here, and would prefer not to use it. However your solution doesn't seem to work on all tests, @ekpyron , so I have to check.

Reply:
Maybe it *is* relevant what prefix these things have...

Reply:
We *could* pull in https://compile-time-regular-expressions.readthedocs.io/en/latest/ for this :-D.
I'm partly kidding, partly serious about that... we do use regular expressions in the optimizer for the name cleanup and I did notice them not being insignificant when profiling...
Here of course it doesn't matter too much performance-wise I guess...

Reply:
That's actually not completely true.